LFS-955: Polish the look of form questions and answers

marta- opened the pull request:
Improved horizontal alignment indentation, vertical spacing, vocabulary input placeholders, answer instructions message. See subtasks of [LFS-955](https://phenotips.atlassian.net/browse/LFS-955) for details.

marta- committed:
LFS-955/LFS-271: When there is no predefined list of options, the input placeholder should be empty instead of displaying "Other (specify)"

marta- committed:
LFS-955/LFS-279: When minAnswers = 1, the message "Please select at least one option" should be replaced with "This question is mandatory"
LFS-955/LFS-280: Improved display of how many answers a user needs to provide to a question

marta- committed:
LFS-955/LFS-956: Answers should not be indented relatively to the question text
LFS-955/LFS-281: Reduce the white space between the question text and the list of options or user input

marta- committed:
LFS-955/LFS-957: Clean up the look of Notes
* Increase the spacing between answers and notes
* The notes textarea should not be shrunk when unfocused

sashaandjic commented:
[] Answers should not be indented relatively to the question text

marta- committed:
LFS-955/LFS-956: Answers should not be indented relatively to the question text
Fixed the indentation of radio button and input for MultipleChoice and Vocabulary questions with maxAnswers=1

marta- committed:
LFS-353: Remove radio button when field is defined as Data Type = vocabulary & MaxAnswers=1
The radio is hidden from the UI. The behavior is still buggy.

marta- committed:
LFS-955/LFS-279: When minAnswers = 1, the message "Please select at least one option" should be replaced with "This question is mandatory"
LFS-955/LFS-280: Improved display of how many answers a user needs to provide to a question
Now displaying answer instructions for all question types

marta- committed:
LFS-955/LFS-956: Answers should not be indented relatively to the question text
Improved markup and simplified styling

marta- committed:
LFS-955/LFS-962: Highlight answer options on hover

sashaandjic reviewed (approved):
tested

sashaandjic reviewed (approved):
tested - looks very good!

sashaandjic commented:
a small bug in the form's view mode

marta- committed:
Merge branch 'dev' into LFS-955

sashaandjic commented:
re-tested after the "dev" branch merge

marta- committed:
LFS-955/LFS-962: Highlight answer options on hover
Removed highlighting in view mode

sashaandjic reviewed (approved):
tested

marta- committed:
LFS-955/LFS-956: Answers should not be indented relatively to the question text
Improved markup and fixed the indentation of radio button and input for Vocabulary questions with maxAnswers=1 (again)

sashaandjic commented:
retested - all good

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -0,0 +1,70 @@
+
+import QuestionnaireStyle from "./QuestionnaireStyle";
+
+// Display instructions regarding how many answers must be provided to a question,
+// based on minAnswers and maxAnswers from the question definition
```
I would make this more generic, we should also add instructions based on minValue/maxValue for numbers for example, or regular expressions/format, or even explicit instructions (what we currently put in "description", but I'm beginning to think it isn't really the right place).

sdumitriu replied:
Hm, cramming multiple sources of instructions into one sounds like a complicated mess. We may need to add a manager/implementations paradigm here in the future.

marta- replied:
We can expand later

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/AnswerInstructions.jsx:
```diff
@@ -0,0 +1,70 @@
+      { (minAnswers == 1 && !(maxAnswers > minAnswers)) ?
+      "This question is mandatory"
+      :
+      "Please provide " + minAnswers + (maxAnswers != minAnswers ? (" - " + maxAnswers) : "" ) + " " + answerLabel + (maxAnswers > 1 ? "s" : "")
```
minAnswers = 3 & maxAnswers = 0 (at least 3 answers) doesn't display properly ("Please provide 3 - 0  value").

sdumitriu replied:
I also think that min==max should be a special case, `3-3 values` doesn't seem right.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -298,20 +299,18 @@ function MultipleChoice(props) {
 
+  // When counting current answers for proper highlighting of answer instructions to the user, exclude the empty one
+  let currentAnswers = answers.filter(item => item[VALUE_POS] !== '').length;
+  const instructions = <AnswerInstructions currentAnswers={currentAnswers} {...props.questionDefinition} {...props} />;
```
If both questionDefinition and props define the same setting, this will lead to duplicate props. Does this work instead?

```suggestion
  const instructions = <AnswerInstructions currentAnswers={currentAnswers} {...props.questionDefinition, ...props} />;
```

marta- replied:
It does not. Duplicate props are not a real problem at this time.

sdumitriu commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -137,15 +137,27 @@ const thesaurusStyle = theme => ({
+  nestedSearchInput: {
+    marginLeft: theme.spacing(4),
+    "& .MuiInputBase-root" : {
+      width: "218px !important",
```
Magic number?

marta- replied:
yes, just like the 250 i'm overriding.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -47,27 +75,28 @@ const questionnaireStyle = theme => ({
         position: 'relative',
     },
+    nestedInput: {
+        minWidth: "218px",
```
Magic number?

marta- replied:
yes, just like the 250 i'm overriding.

k-pace commented:
Tested, love the improvements! No bugs found from me

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -73,15 +74,20 @@ function VocabularyQuestion(props) {
     // Reparse defaults into a format VocabularySelector understands
     .reduce((object, value) => ({...object, [value[VALUE_POS]]: value[LABEL_POS]}), {});
 
+  let [currentAnswers, setCurrentAnswers] = useState(props.defaults?.length || 0);
```
Are you sure about this? `defaults` is not a list of _selected_ answers, just predefined options to choose from. And it never seems to be passed to a VocabularyQuestion.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -73,15 +74,20 @@ function VocabularyQuestion(props) {
+      disableInstructions
+      {...props}
       >
+      <AnswerInstructions currentAnswers={currentAnswers} {...props.questionDefinition} />
```
I don't like that we have to duplicate this bit of code in different places. Makes adding more functionality to `AnswerInstructions` harder... Not something that can be easily done in this PR, just signaling a code smell.

sdumitriu reviewed (changes requested):
It seems to me that there are too many changes in the vocabulary code. Indicates that it may benefit from a code cleanup in the near future.

marta- committed:
LFS-955/LFS-280: Improved display of how many answers a user needs to provide to a question
Adressing review comments: Fixed the logic behind message announcing how many values should be entered

marta- committed:
LFS-955/LFS-280: Improved display of how many answers a user needs to provide to a question
Adressing review comments: Fixed the initialization of number of values provided for vocabulary questions

marta- committed:
LFS-955/LFS-280: Improved display of how many answers a user needs to provide to a question
Adressing review comments: Removed unnecessary override of answer instructions in vocabulary questions

marta- committed:
Merge branch 'dev' into LFS-955

sdumitriu reviewed (dismissed):
Regression: vocabulary questions accept max+1 answers before disabling other inputs.
- set maxAnswers=3 for Demographics/Ethnicity
- create a new Demographics forms
- select 4 ethnicities
Works correctly for non-vocabulary lists, and works well in dev.

sdumitriu commented:
Ah, nevermind, it's an old bug, triggered by how many options were already selected before re-editing.

sdumitriu commented:
Weird corner case (that should never happen in practice): minAnswers=3, maxAnswers=2, forces me to select 3 answers, but the controls get disabled after only 2 are selected. Should we enforce maxAnswers = max(maxAnswers, minAnswers)?

sashaandjic commented:
retested in 0.8.0 smoke test
